### DISCUSSION TITLE
How can I make sure my WSL2/g distro is using vGPU?

### DISCUSSION THREAD

**Discussion by ghost** [microsoft/wslg#78] [Q&A] [answered]
in [this page ](https://devblogs.microsoft.com/commandline/wslg-architecture/)there are links to Intel, Nvidia, AMD driver download pages. I'm currently using Intel built in GPU, the link to that driver in that page is old. I'm on latest build of Windows 10 Dev, Windows update automatically installs the latest Intel GPU driver.
so I don't know whether I should uninstall the new version of the driver and install that old one, or I should do something else to activate vGPU.

can someone help me with this?
I installed Edge Dev on Ubuntu, it doesn't have hardware acceleration support, so I think vGPU is not activated.

How do you check for the availability of vGPU for WSLg in the Host and Guest OS ?

**Comment by onomatopellan** [comment]
First of all make sure the GPU appears inside of wsl2. If the file `/dev/dxg` exists you have the correct driver.
For GPU acceleration on GUI you will need to install latest Mesa with d3d12 support:
```
sudo add-apt-repository ppa:kisak/kisak-mesa
sudo apt-get update
```

After that run something like `glxinfo` and in the OpenGL Renderer part it should say D3D12 and your GPU name.

**Reply by ghost** [reply]
/dev/dxg file exists but empty.
i also did the commnads, not sure what to look for in glxinfo, lots of texts appear.

**Reply by onomatopellan** [reply]
What's the output of `glxinfo | grep -i opengl` ?

**Reply by badjames** [reply]
That only adds the repository and updates the available packages.

You also need to actually upgrade mesa using "sudo apt install".

I'm not sure what the best command is, but this worked for me:

sudo apt install libegl-mesa0

After running that, my glxinfo renderer switched from the dreaded llvmpipe software renderer to the correct NVIDIA (in my case) GPU.

**Comment by crramirez** [comment]
That said, even though you have vGPU correctly configured and can see your GPU in glxinfo, you won't see acceleration in Edge. Why? I don't know.

**Reply by onomatopellan** [reply]
> That said, even though you have vGPU correctly configured and can see your GPU in glxinfo, you won't see acceleration in Edge. Why? I don't know.

I finally got hardware acceleration for my GT 710 on Edge-dev. I had to compile and install latest mesa's main branch (version 21.2.0-devel).

Edit: Actually it was activating "Override software rendering list" what made me see GPU acceleration in WSL2.

**Comment by spronovo** [comment]
Could you share the output of your ```glxinfo -B``` you should see something like below where you can see that i'm running Mesa 21.x and the device in my case is D3D12 (NVIDIA GeForce RTX 3090), if the renderer is not D3D12, you won't be getting hardware acceleration.

```
spronovo@OFFICE:~$ glxinfo -B
name of display: :0
NVD3D10: CPU cyclestats are disabled on client virtualization
NVD3D10: CPU cyclestats are disabled on client virtualization
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Microsoft Corporation (0xffffffff)
    Device: D3D12 (NVIDIA GeForce RTX 3090) (0xffffffff)
    Version: 21.1.0
    Accelerated: yes
    Video memory: 89855MB
    Unified memory: no
    Preferred profile: core (0x1)
    Max core profile version: 3.3
    Max compat profile version: 3.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.0
OpenGL vendor string: Microsoft Corporation
OpenGL renderer string: D3D12 (NVIDIA GeForce RTX 3090)
OpenGL core profile version string: 3.3 (Core Profile) Mesa 21.1.0-devel (git-695d3bcb12)
OpenGL core profile shading language version string: 3.30
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 3.1 Mesa 21.1.0-devel (git-695d3bcb12)
OpenGL shading language version string: 1.40
OpenGL context flags: (none)

OpenGL ES profile version string: OpenGL ES 3.0 Mesa 21.1.0-devel (git-695d3bcb12)
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.00
```

For Edge, are you using ```edge://gpu``` to look at your GPU status? What is it showing?

![image](https://user-images.githubusercontent.com/7545376/116353800-3deba900-a7ac-11eb-919a-9876efb2b3dd.png)

**Reply by crramirez** [reply]
I cannot get acceleration in Edge but in other apps like glxinfo. Any other thing I need to do?

Another thing I see that you are using Mesa 21.1 instead of 21.0 do you recommend it?

**Comment by spronovo** [comment, marked as answer]
> in [this page ](https://devblogs.microsoft.com/commandline/wslg-architecture/)there are links to Intel, Nvidia, AMD driver download pages. I'm currently using Intel built in GPU, the link to that driver in that page is old. I'm on latest build of Windows 10 Dev, Windows update automatically installs the latest Intel GPU driver.
> so I don't know whether I should uninstall the new version of the driver and install that old one, or I should do something else to activate vGPU.
> 
> can someone help me with this?
> I installed Edge Dev on Ubuntu, it doesn't have hardware acceleration support, so I think vGPU is not activated.
> 
> Do do you check for the availability of vGPU for WSL in the Host and Guest OS ?

For Intel you actually do need to use that old driver at the moment as it is the only one that exposes vGPU support in WSL, the currently flighted driver on WU doesn't expose this yet, but this is something we're working with Intel on addressing.

Verify that you are running on a the recently release 5.10 kernel

```
spronovo@OFFICE:~$ uname -a
Linux OFFICE 5.10.16.3-microsoft-standard-WSL2 #1 SMP Fri Apr 2 22:23:49 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

Look for the existence of /dev/dxg. If it exist, it means we've detected at least one vGPU capable device. Note that on older kernel (5.4) /dev/dxg would be created even when no actual vGPU was found... but on 5.10+ it is only created if we've actually enumerated a device over vPCI... so it's a quick way of determining if you have a vGPU.

**Reply by ghost** [reply]
Hi,
thank you for confirmation and all the information, I wasn't aware of that, so I thought the only requirement is to have WDDM 3 on host.
but now knowing I only need that old driver, I think I will wait until the new drivers start supporting WSLg too and looking forward to that. thank you

**Reply by Arctic-Xiangjian** [reply]
Hello, I update the edge today, the output for `glxinfo -B` changed. Now it become
```name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Mesa/X.org (0xffffffff)
    Device: llvmpipe (LLVM 12.0.0, 256 bits) (0xffffffff)
    Version: 21.0.3
    Accelerated: no
    Video memory: 15967MB
    Unified memory: no
    Preferred profile: core (0x1)
    Max core profile version: 4.5
    Max compat profile version: 3.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.2
OpenGL vendor string: Mesa/X.org
OpenGL renderer string: llvmpipe (LLVM 12.0.0, 256 bits)
OpenGL core profile version string: 4.5 (Core Profile) Mesa 21.0.3 - kisak-mesa PPA
OpenGL core profile shading language version string: 4.50
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 3.1 Mesa 21.0.3 - kisak-mesa PPA
OpenGL shading language version string: 1.40
OpenGL context flags: (none)

OpenGL ES profile version string: OpenGL ES 3.2 Mesa 21.0.3 - kisak-mesa PPA
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
```


But the last version it could work rightly. My gpu is RTX2070

**Reply by spronovo** [reply]
If you picked an updated build of Windows today, your graphics driver might have gotten reset and you may need to manually reinstall the WSL enabled driver. You can launch ```dxdiag.exe``` from Windows to look at your driver version

![image](https://user-images.githubusercontent.com/7545376/116774671-0d597880-aa13-11eb-845b-94af7a4add84.png)

**Reply by Arctic-Xiangjian** [reply]
Thank you very much~
It really is the driver problems.

**Comment by aki-k** [comment]
Here are my notes on how I was able to get the D3D12 device shown in glxinfo -B in WSL 2/WSLg:

https://pastebin.com/raw/f0SpM3C1

**Comment by Z-X20A** [comment]
Hi, I try all the way mentioned above. It still not work. I try sudo apt install libegl-mesa0 
![image](https://github.com/user-attachments/assets/2248252a-a089-4d34-b5e6-3a37d68e4497)

**Reply by onomatopellan** [reply]
@Z-X20A What's your GPU? Do you have the correct Windows driver version installed? What do you see if you run `ls -l -a /usr/lib/wsl/lib` in the terminal?

**Reply by Z-X20A** [reply]
Hi, my gpu is rtx 3070 laptop. after running, it show.
![image](https://github.com/user-attachments/assets/7ee2cdbe-1475-471a-9d12-9d503df3f3c9)

**Reply by onomatopellan** [reply]
See https://github.com/microsoft/wslg/issues/40#issuecomment-1685123693

If you have latest Windwos Driver, latest mesa and latest WSL (`wsl.exe --version`) it should work.

Edit: Also try in a new clean distro you just installed like "Ubuntu 24.04 LTS". It should just work without installing anything.

**Reply by thirumurug0xan** [reply]
Does it work with Intel igpu if install ubuntu 24?
```
┌──(kaligpu3㉿Thirumurugan)-[~]
└─$ ls -l -a /usr/lib/wsl/lib
total 8428
drwxr-xr-x 1 root root      40 Nov  3 16:47 .
drwxr-xr-x 4 root root    4096 Nov  1 23:00 ..
-r-xr-xr-x 1 root root 6880344 Oct 20  2023 libd3d12core.so
-r-xr-xr-x 1 root root  801840 Oct 20  2023 libd3d12.so
-r-xr-xr-x 1 root root  942048 Mar 31  2024 libdxcore.so_ 
```
